### DISCUSSION TITLE
I/V curves: pvlib.pvsystem.calcparams_pvsyst vs PVsyst 7.2

### DISCUSSION THREAD

**Discussion by FrancescoMariottini** [pvlib/pvlib-python#1783] [Q&A]
Hello everybody,

I would like to kindly ask you to share your testing experience with the function pvlib.pvsystem.calcparams_pvsyst against PVsyst.

From my experience I/V curves based on the function pvlib.pvsystem.calcparams_desoto are actually way closer to the  I/V curves showed in PVsyst (section "Definition of a PV module") than curves based on pvlib.pvsystem.calcparam.

The parameters I am using in pvlib correspond to the ones defined for the module in PVsyst (SolarWorld 250 Wp 26 V, Si-poly):
{'alpha_sc': 0.0029,  #muIsc  (often named Alpha)
 'gamma_ref': 0.94, # The diode ideality factor # muIsc  (often named Alpha)
 'mu_gamma': 0.001, # The temperature coefficient for the diode ideality factor, 1/K
 'I_L_ref': 8.64, #  Isc
 'I_o_ref': 4.5999999999999996e-11, #  The dark or diode reverse saturation current at reference conditions, in amperes.
 'R_sh_ref': 300, # The shunt resistance at reference conditions, in ohms.
 'R_sh_0': 1200, # The shunt resistance at zero irradiance conditions, in ohms.
 'R_s': 0.3, #  series model
 'cells_in_series': 60, # 
 'R_sh_exp': 5.5, # The exponent in the equation for shunt resistance, unitless. Defaults to 5.5.
 'EgRef': 1.12, # The energy bandgap at reference temperature in units of eV. 1.121 eV for crystalline silicon. EgRef must be >0.
 'irrad_ref': 1000, 
 'temp_ref': 25 
}

I exclude that the gap between pvlib.pvsystem.calcparams_pvsyst and pvsyst is due to the fact that function is based on the version 6 of pvsyst ([link](https://pvlib-python.readthedocs.io/en/stable/reference/generated/pvlib.pvsystem.calcparams_pvsyst.html?highlight=calcparam#pvlib.pvsystem.calcparams_pvsyst)) since I haven't found any sensible difference in PVsyst I/V curves representation between the version 5 and 7.2 (tested on Generic Poly 110 Wp 1x72 cells).

Thanks in advance

[
![iv_curves_pvlib_vs_pvsyst](https://github.com/pvlib/pvlib-python/assets/70909009/26849f3c-8612-496d-b706-b74444ff28e3)
](url)

**Comment by cwhanse** [comment]
@FrancescoMariottini Can you clarify the question a bit? Are you saying that the left plots (pvlib, desoto model with parameters from ?) are closer to the center plots (are these from PVsyst?) than are the right plots (pvlib, pvsyst model with parameters from PVsyst)?

If that's what you are asking, could you post the Desoto model parameters, and could you provide the MPP values for all three sets of curves?

**Reply by FrancescoMariottini** [reply]
Hello @cwhanse that's exactly what I mean. Thanks for your fast reply.

You will find attached the output csvs.  Hereby the used code (Jupyter Notebook saved as Python code)

[iv_curves_pvsyst_i1000.csv](https://github.com/pvlib/pvlib-python/files/11880374/iv_curves_pvsyst_i1000.csv)
[iv_curves_pvsyst_t45.csv](https://github.com/pvlib/pvlib-python/files/11880375/iv_curves_pvsyst_t45.csv)
[iv_curves_desoto_i1000.csv](https://github.com/pvlib/pvlib-python/files/11880376/iv_curves_desoto_i1000.csv)
[iv_curves_desoto_t45.csv](https://github.com/pvlib/pvlib-python/files/11880377/iv_curves_desoto_t45.csv)


# %% [markdown]
# # Comparison pvlib vs pvsyst
# 
# ## Libraries import

# %%

```
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import pvlib
```

# %% [markdown]
# ## Global variables

# %%
```
k_Boltzmann = 1.381 / (10 ** 23) # J/K
q_electron = 1.602 / (10 ** 19) # C


pvsyst_SolarWorld_Sunmodule_250_Poly__2013 = {
#The short-circuit current temperature coefficient of the module in units of A/C.
# muIsc  (often named Alpha)
"alpha_sc" : 2.9 / (10**3), 
#The diode ideality factor 
"gamma_ref" : 0.94, 
# The temperature coefficient for the diode ideality factor, 1/K
# using 2nd value comparing in the diode ideality factor 
"mu_gamma" : 0.001, 
#The light-generated current (or photocurrent) at reference conditions, in amperes.
# using Isc for I_L_ref
"I_L_ref" : 8.640, 
#The dark or diode reverse saturation current at reference conditions, in amperes.
"I_o_ref" : 0.046 / (10 ** 9), 
#The shunt resistance at reference conditions, in ohms.
"R_sh_ref" : 300, 
#The shunt resistance at zero irradiance conditions, in ohms.
"R_sh_0" : 1200, 
#The series resistance at reference conditions, in ohms.
# series model
"R_s" : 0.3, 
#The number of cells connected in series.
"cells_in_series" : 60, 
#The exponent in the equation for shunt resistance, unitless. Defaults to 5.5.
# PVsyst: this value is set as default value ... Navigation:  Physical models used > PV Module - Standard one-diode-model >
"R_sh_exp" : 5.5, 
#The energy bandgap at reference temperature in units of eV. 1.121 eV for crystalline silicon. EgRef must be >0.
"EgRef" : 1.12, #1, #tried also with suggested 1.121 , no sensible changes
# Reference irradiance in W/m^2.
"irrad_ref" : 1000, 
"temp_ref" : 25
}

# %%
ps_prms = ps_prms = pvsyst_SolarWorld_Sunmodule_250_Poly__2013
```

# %% [markdown]
# ## Wrap-up functions

# %%
```
def get_iv(df:pd.DataFrame, ps_prms:dict, calcparams="desoto", ivcurve_pnts=400) -> pd.DataFrame:
    """
    wrap-up function
    :param df:
    :param x: dataframe dictionary
    :param ps_prms: parameters dictionary 
    :param calcparams: "desoto" or "pvsyst"
    :param calcmp: "singlediode", "bishop88"
    :param method: 'lambertw' #, 'newton', or 'brentq'
    :param ivcurve_pnts: 
    :return: df with additional columns (see below). i & v as array for each point.
    """ 
    # using the ones from pvsyst instead of pvlib
    temp_cell = "TArray"
    effective_irradiance = "GlobEff"
    # no need to order
    clms_sd = ["photocurrent","saturation_current","resistance_series","resistance_shunt","nNsVth"]
    clms = ["i_sc","v_oc","i_mp","v_mp","p_mp","i_x","i_xx"]
    if ivcurve_pnts > 0: clms = clms + ["v","i"]

    if calcparams == "desoto":
        df[clms_sd] = df.apply(
        lambda x:                 
        pvlib.pvsystem.calcparams_desoto(
        effective_irradiance = x[effective_irradiance], 
        temp_cell = x[temp_cell], 
        alpha_sc = ps_prms["alpha_sc"], #The short-circuit current temperature coefficient of the module in units of A/C.
        a_ref =  ps_prms["gamma_ref"] * ps_prms["cells_in_series"] * k_Boltzmann * 298.16 / q_electron,
        # The product of the usual diode ideality factor (n, unitless), number of cells in series (Ns), and cell thermal voltage at reference conditions, in units of V        
        # {\displaystyle V_{\text{T}}=kT/q,} the thermal voltage.
        # gamma_ref = ps_prms["gamma_ref"], #The diode ideality factor
        # mu_gamma = ps_prms["mu_gamma"], #The temperature coefficient for the diode ideality factor, 1/K
        I_L_ref = ps_prms["I_L_ref"], #The light-generated current (or photocurrent) at reference conditions, in amperes.
        I_o_ref = ps_prms["I_o_ref"], #The dark or diode reverse saturation current at reference conditions, in amperes.
        R_sh_ref = ps_prms["R_sh_ref"], #The shunt resistance at reference conditions, in ohms.
        # R_sh_0 = ps_prms["R_sh_0"], #The shunt resistance at zero irradiance conditions, in ohms.
        R_s = ps_prms["R_s"], #The series resistance at reference conditions, in ohms.
        # cells_in_series = ps_prms["cells_in_series"], #The number of cells connected in series.
        #R_sh_exp=ps_prms["R_sh_exp"], #The exponent in the equation for shunt resistance, unitless. Defaults to 5.5.
        dEgdT= - 0.0002677, #The temperature dependence of the energy bandgap at reference conditions in units of 1/K
        EgRef=ps_prms["EgRef"], #The energy bandgap at reference temperature in units of eV. 1.121 eV for crystalline silicon. EgRef must be >0.
        irrad_ref=ps_prms["irrad_ref"], # Reference irradiance in W/m^2.
        temp_ref=ps_prms["temp_ref"] # Reference cell temperature in C.
        ) if x[effective_irradiance] != 0 else
        pd.Series(dict(zip(clms_sd,[np.nan]*len(clms_sd)))), 
        axis=1, result_type='expand')

    # version 0.93, PVsyst v6 model
    elif calcparams == "pvsyst":
        df[clms_sd] = df.apply(lambda x:                 
        pvlib.pvsystem.calcparams_pvsyst(
        effective_irradiance = x[effective_irradiance], 
        temp_cell = x[temp_cell], 
        alpha_sc = ps_prms["alpha_sc"], #The short-circuit current temperature coefficient of the module in units of A/C.
        gamma_ref = ps_prms["gamma_ref"], #The diode ideality factor
        mu_gamma = ps_prms["mu_gamma"], #The temperature coefficient for the diode ideality factor, 1/K
        I_L_ref = ps_prms["I_L_ref"], #The light-generated current (or photocurrent) at reference conditions, in amperes.
        I_o_ref = ps_prms["I_o_ref"], #The dark or diode reverse saturation current at reference conditions, in amperes.
        R_sh_ref = ps_prms["R_sh_ref"], #The shunt resistance at reference conditions, in ohms.
        R_sh_0 = ps_prms["R_sh_0"], #The shunt resistance at zero irradiance conditions, in ohms.
        R_s = ps_prms["R_s"], #The series resistance at reference conditions, in ohms.
        cells_in_series = ps_prms["cells_in_series"], #The number of cells connected in series.
        R_sh_exp=ps_prms["R_sh_exp"], #The exponent in the equation for shunt resistance, unitless. Defaults to 5.5.
        EgRef=ps_prms["EgRef"], #The energy bandgap at reference temperature in units of eV. 1.121 eV for crystalline silicon. EgRef must be >0.
        irrad_ref=ps_prms["irrad_ref"], # Reference irradiance in W/m^2.
        temp_ref=ps_prms["temp_ref"] # Reference cell temperature in C.
        ) if x[effective_irradiance] != 0 else
        pd.Series(dict(zip(clms_sd,[np.nan]*len(clms_sd)))), 
        axis=1, result_type='expand')


    df[clms] = df.apply(lambda x:                 
    pvlib.pvsystem.singlediode(
    photocurrent = x["photocurrent"],
    saturation_current = x["saturation_current"],
    resistance_series = x["resistance_series"],
    resistance_shunt = x["resistance_shunt"],
    nNsVth = x["nNsVth"], # k_Boltzmann * x[temp_cell] / q_electron
    ivcurve_pnts = ivcurve_pnts,
    method = "lambertw"
    ) if x[effective_irradiance] != 0 else
    pd.Series(dict(zip(clms,[np.nan]*len(clms))))
    , axis=1, result_type='expand')
    # not clear why array returned
    df["i_mp"]=df["i_mp"].apply(float)
    df["v_mp"]=df["v_mp"].apply(float)

    return df
```

# %%
```
def plot_iv_curves(irradiance: list, temperature:list, calcparams = "desoto") -> pd.DataFrame:
    plt.close()
    fig,ax1 = plt.subplots()
    calccmp ="singlediode"
    title = f"IV curves (parameters {calcparams}, mpp {calccmp})"
    iv_conditions = pd.DataFrame({"GlobEff":irradiance,"TArray":temperature})
    iv_curves = get_iv(iv_conditions, ps_prms, calcparams)         
    for k, r in iv_curves.iterrows():
        l = "_".join(["GlobEff",str(r["GlobEff"]),"TArray",str(r["TArray"])])
        ax1.plot(r["v"], r["i"], label=l, alpha=0.7)
    ax1.set_title(title)
    ax1.set_xlabel('v')
    ax1.set_ylabel('i')
    ax1.legend()
    return iv_curves
```

# %% [markdown]
# ## I/V curves plotting

# %%
```
calcparamss = ["desoto", "pvsyst"]
calccmp = "singlediode"
```

# %%
```
iv_curves = plot_iv_curves(irradiance = [1000,800,600,400,200], temperature = [45]*5,
                            calcparams="pvsyst")
iv_curves.to_csv("iv_curves_pvsyst_t45.csv")
```

# %%
```
iv_curves = plot_iv_curves(irradiance = [1000,800,600,400,200], temperature = [45]*5,
                            calcparams="desoto")
iv_curves.to_csv("iv_curves_desoto_t45.csv")
```

# %%
```
iv_curves = plot_iv_curves(irradiance = [1000]*5, temperature = [10,25,40,55,70],
                            calcparams="pvsyst")
iv_curves.to_csv("iv_curves_pvsyst_i1000.csv")
```

# %%
```
iv_curves = plot_iv_curves(irradiance = [1000]*5, temperature = [10,25,40,55,70],
                            calcparams="desoto")
iv_curves.to_csv("iv_curves_desoto_i1000.csv")
```

**Comment by cwhanse** [comment]
It appears that you are assuming that the Desoto model parameters (e.g. R_sh_ref) are the same as the parameters in the PVsyst model. Although the parameters have the same (or similar) physical interpretation, mathematically they aren't equivalent: each set of parameters results from fitting a different set of equations to the (same) data, so we can't regard two parameters with the same name as equivalent. Also, the fitting method matters - two different methods to fit the same set of equations will often yield different parameters.

I prefer to think of the parameter set as intimately connected with both the model (e.g. PVsyst) and the method used to fit that model.

I note that the SolarWorld module is in the CEC database

```
modules = pvlib.pvsystem.retrieve_sam('CECmod')
swmod = modules['SolarWorld_Industries_GmbH_Sunmodule_Plus_SW_250_poly']
ivp = pvlib.pvsystem.calcparams_cec(np.array([1000]*5), np.array([10,25,40,55,70]),
                                    swmod['alpha_sc'], swmod['a_ref'], swmod['I_L_ref'],
                                    swmod['I_o_ref'], swmod['R_sh_ref'], swmod['R_s'],
                                    swmod['Adjust'])
ivs = pvlib.pvsystem.singlediode(*ivp)
```

Because the Adjust parameter is small (1.4%) you can use these same parameters for the Desoto model also.

By inspection, the CEC model results follow the PVsyst values more closely than does pvlib's implementation of the PVsyst equations. Note that the parameters are different than the PVsyst values, in particular the shunt resistance.

I don't have a PVsyst license so I can't verify that PVsyst reproduces the values in the figures you posted - I'm assuming that you have done that.

What does all this mean? My view is that it is not so much the model that matters, but how its parameters are determined.

Why do the results from pvlib's implementation of the PVsyst equations vary from the PVsyst figures you display? I don't know. The pvlib implementation is based on the best information we have about what PVsyst does. It is against the PVsyst terms of use to reverse engineer their software. It is possible (maybe likely) that pvlib doesn't faithfully imitate PVsyst, but I don't know how we can improve pvlib in this regard.

**Reply by FrancescoMariottini** [reply]
Thanks again for taking the time to reply.

> Because the Adjust parameter is small (1.4%) you can use these same parameters for the Desoto model also.
> 
> -- <cite>Cliff Hansen</cite>

I see no gamma_ref in CEC 'retrieve_sam'. Could you please tell me which value did you use to apply the function "calcparams_desoto" ?

> By inspection, the CEC model results follow the PVsyst values more closely than does pvlib's implementation of the PVsyst equations.
> 
> -- <cite>Cliff Hansen</cite>

The module SolarWorld_Industries_GmbH_Sunmodule_Plus_SW_250_poly available in CEC (and pvsyst) is different than the module pvsyst_SolarWorld_Sunmodule_250_Poly__2013 I have previously showed (available in pvsyst but not in CEC). Hereby the pvsyst parameters for the module you used:

`
pvsyst_SolarWorld_Industries_GmbH_Sunmodule_Plus_SW_250_poly = {
"alpha_sc" : 4.4 / (10**3), # muIsc  (often named Alpha)
"gamma_ref" : 0.961,  #The diode ideality factor 
"mu_gamma" : 0.00, #04 # replaced 0 (?!!) # The temperature coefficient for the diode ideality factor, 1/K
"I_L_ref" : 8.810, # using Isc 
"I_o_ref" : 0.050 / (10 ** 9), #The dark or diode reverse saturation current at reference conditions, in amperes.
"R_sh_ref" : 240, #The shunt resistance at reference conditions, in ohms.
"R_sh_0" : 1000, #The shunt resistance at zero irradiance conditions, in ohms.
"R_s" : 0.35, #The series resistance at reference conditions, in ohms.
 #same as pvsyst_SolarWorld_Sunmodule_250_Poly__2013 :
"cells_in_series" : 60, 
"R_sh_exp" : 5.5, # The exponent in the equation for shunt resistance
"EgRef" : 1.12, # The energy bandgap at reference temperature in units of eV. 1.121 eV
"irrad_ref" : 1000,  # Reference irradiance in W/m^2.
"temp_ref" : 25    
}
`

When comparing the three methods (see the attached jpg and csvs), the Desoto is still closer to pysyst than the pvsyst method itself. The I/V curves at Gref 1000 W/m2 for Desoto are also the closest to the ones in pvsyst. CEC method is closest ones when considering I/V curves at 45° C.

![image](https://github.com/pvlib/pvlib-python/assets/70909009/a2225fae-8d8c-4657-8780-706135053d1e)

[iv_curves_desoto_i1000.csv](https://github.com/pvlib/pvlib-python/files/11915000/iv_curves_desoto_i1000.csv)
[iv_curves_desoto_t45.csv](https://github.com/pvlib/pvlib-python/files/11915003/iv_curves_desoto_t45.csv)
[iv_curves_pvsyst_i1000.csv](https://github.com/pvlib/pvlib-python/files/11915004/iv_curves_pvsyst_i1000.csv)
[iv_curves_pvsyst_t45.csv](https://github.com/pvlib/pvlib-python/files/11915005/iv_curves_pvsyst_t45.csv)
[iv_curves_cecmod_i1000.csv](https://github.com/pvlib/pvlib-python/files/11915006/iv_curves_cecmod_i1000.csv)
[iv_curves_cecmod_t45.csv](https://github.com/pvlib/pvlib-python/files/11915007/iv_curves_cecmod_t45.csv)

I personally found the "Model parameters" section of "Sunmodule Plus SW 250 poly" in PVsyst particularly problematic. 
On the line "Diode quality factor", Gamma is defined as "0.96 - 0.000/K".
I generally consider the first parameter (0.96) as gamma_ref, but in this specific case I found very strange to use the second value (0.000) as a null temperature coefficient for the diode ideality factor mu_gamma (maybe an approximation/visualisation error ?). 

> I don't have a PVsyst license so I can't verify that PVsyst reproduces the values in the figures you posted.
> 
> -- <cite>Cliff Hansen</cite>

pvlib "calcparams_pvsyst" method could be compared with the I/V curves of the generic pv modules available also for the trial version of pvsyst. There is no need for a license.
**I would gladly reproduce the previous results for another pv module of your choice in case you would like to have a deeper look at it.**

> My view is that it is not so much the model that matters, but how its parameters are determined.
> 
> -- <cite>Cliff Hansen</cite>

I agree with you on the importance of parameters fitting. 

Still I find very difficult to trust and use the method "calcparams_pvsyst" when the results seem considerable different than pvsyst while other methods (desoto and CEC) seem better.

And I think that is a pity to validate the modeling against pvsyst using a different method (desoto, rather than CEC since my target module is not available) rather than pvlib.pvsystem.calcparams_pvsyst.

**Reply by cwhanse** [reply]
The CEC model parameters provide 'a_ref', which is gamma_ref * cells in series * thermal voltage (k/q * cell temperature in K).

I agree, it's a pity that we are uncertain about what exactly is the diode model in PVsyst. Could the disparity come from the methods used to create the PAN file you have? From what I've heard, PAN file creation is mostly a black box, containing some optimization procedure. The results depend both on the data provided and on the optimization objectives.

**Comment by FrancescoMariottini** [comment]
> The CEC model parameters provide 'a_ref', which is gamma_ref * cells in series * thermal voltage (k/q * cell temperature in K
> 
> -- <cite>Cliff Hansen</cite>

Sorry, what I actually need is the Adjust parameter since I have already previously calculated a_ref for De Soto model.

 > Could the disparity come from the methods used to create the PAN file you have? 
> 
> -- <cite>Cliff Hansen</cite>

Thanks for the suggestion, I will have another look at the PAN files.
PAN files are provided by manufacturers and their content seem quite transparent to me (which don't exclude the possibility of errors). See also: [Modules in the Database](https://www.pvsyst.com/help/pvmodule_database.htm).
